gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -31,120 +31,148 @@
+                workUnitState.getProp(writerFilePathKey));
+
+        if (writerOutputPathsMoved.contains(writerOutputDir)) {
+          // This writer output path has already been moved for another task of the same extract, so skip to the next one
           continue;
         }
 
-        // Create the parent directory of the final output directory if it does not exist
-        if (!this.fss.get(i).exists(publisherOutput.getParent())) {
-          this.fss.get(i).mkdirs(publisherOutput.getParent());
-        }
+        if (this.fss.get(i).exists(publisherOutputDir)) {
+
+          // The final output directory already exists, check if the job is configured to replace it
+          boolean replaceFinalOutputDir =
+              this.getState().getPropAsBoolean(
+                  ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.DATA_PUBLISHER_REPLACE_FINAL_DIR,
+                      branches, i));
 
-        if (this.fss.get(i).exists(publisherOutput)) {
-          // The final output directory already exists, check if configured to replace it.
-          boolean replaceFinalOutputDir = this.getState().getPropAsBoolean(ForkOperatorUtils
-                  .getPropertyNameForBranch(ConfigurationKeys.DATA_PUBLISHER_REPLACE_FINAL_DIR, branches, i));
+          // If the final output directory is not configured to be replaced, then append the new data to the existing
+          // output folder
           if (!replaceFinalOutputDir) {
-            // The final output directory is not configured to be replaced,
-            // add the output files to the existing final output directory.
-            // TODO: revisit this part when a use case arises
-            boolean preserveFileName = workUnitState.getPropAsBoolean(ForkOperatorUtils
-                    .getPropertyNameForBranch(ConfigurationKeys.SOURCE_FILEBASED_PRESERVE_FILE_NAME, branches, i),
-                false);
-            for (FileStatus status : this.fss.get(i).listStatus(writerOutput)) {
-              // Preserve the file name if configured, use specified name otherwise.
-              Path outputPath = preserveFileName ? new Path(publisherOutput, workUnitState.getProp(
-                  ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.DATA_PUBLISHER_FINAL_NAME, branches, i)))
-                  : new Path(publisherOutput, status.getPath().getName());
-              if (this.fss.get(i).rename(status.getPath(), outputPath)) {
-                LOG.info(String.format("Moved %s to %s", status.getPath(), outputPath));
+
+            boolean preserveFileName =
+                workUnitState.getPropAsBoolean(ForkOperatorUtils.getPropertyNameForBranch(
+                    ConfigurationKeys.SOURCE_FILEBASED_PRESERVE_FILE_NAME, branches, i), false);
+
+            // Go through each file in writerOutputDir and move it into publisherOutputDir
+            for (FileStatus status : this.fss.get(i).listStatus(writerOutputDir)) {
```

Comment:
Yeah, the current version of the `BaseDataPublisher` won't work. I believe your new implementation of `DataPublisher` in https://github.com/linkedin/gobblin/pull/83 will work.